cmd/activator/round_trippers.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2018 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	h2cutil "github.com/knative/serving/pkg/h2c"
+	"go.uber.org/zap"
+)
+
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+
+func (rt roundTripperFunc) RoundTrip(r *http.Request) (*http.Response, error) {
+	return rt(r)
+}
+
+// baseTransport will use the appropriate transport for the request's http protocol version
+var baseTransport http.RoundTripper = roundTripperFunc(func(r *http.Request) (*http.Response, error) {
+	var transport http.RoundTripper = http.DefaultTransport
+	if r.ProtoMajor == 2 {
+		transport = h2cutil.DefaultTransport
```

Comment:
My preference would be getting rid of this func and providing http1 and http2 transports as inputs to the rountripper. That way, we will get better coverage (i.e. the logic to select rounttripper is also covered) and also the tests wouldn't have to know the internals of rounttripper as much.

See: https://github.com/mdemirhan/serving/blob/2bf10f2fca8e56ea174b3259287f62bc2385f26c/pkg/activator/httptransport.go#L33 as an example. 

And corresponding fake transport: https://github.com/mdemirhan/serving/blob/2bf10f2fca8e56ea174b3259287f62bc2385f26c/pkg/activator/httptransport_test.go#L38

And table tests:
https://github.com/mdemirhan/serving/blob/2bf10f2fca8e56ea174b3259287f62bc2385f26c/pkg/activator/httptransport_test.go#L69

Reply:
It feels a little weird to have `RetryRoundTripper` be concerned with transport protocol selection. If there was better test coverage, would you be OK with keeping this logic separate?

Reply:
Yes, that sounds good to me. Thanks!